dali/pipeline/executor/executor2/exec_graph.cc

```diff
@@ -0,0 +1,293 @@
+  }
+}
+
+void ExecGraph::PrepareIteration(const WorkspaceParams &params) {
+  if (nodes_.empty())
+    throw std::logic_error("Cannot execute an empty graph");
+
+  if (params.batch_size <= 0)
+    throw std::runtime_error("Batch size must be a positive number.");
+
+  Sort();
+  Validate();
+  Analyze();
+
+  // Create a special task that checks the predicted batch size
+  // and populates the respective field in IterationData
+  auto prev_infer = infer_batch_size_task_;
+  infer_batch_size_task_ = InferBatchSize(params.iter_data, params.batch_size);
+  if (infer_batch_size_task_) {
+    if (prev_infer)
+      infer_batch_size_task_->Succeed(prev_infer);
+  } else {
+    params.iter_data->default_batch_size = params.batch_size;
+  }
+
+  for (auto &n : nodes_) {
+    n.NextIter();
+    n.CreateMainTask(params);
+  }
+  for (auto &n : nodes_) {
+    n.AddDataDeps();
+    n.CreateAuxTasks();
+    // all root tasks must succeed batch size inference
+    if (n.inputs.empty() && infer_batch_size_task_)
+      n.main_task_->Subscribe(infer_batch_size_task_);
+  }
+}
+
+inline std::string_view NodeName(const ExecNode &n) {
+  return !n.instance_name.empty() ? n.instance_name : n.op->GetSpec().SchemaName();
+}
+
+tasking::SharedTask ExecGraph::InferBatchSize(const std::shared_ptr<IterationData> &iter_data,
+                                              int max_batch_size) {
+  std::vector<std::pair<ExecNode *, BatchSizeProvider *>> bsps;
+  for (auto &n : nodes_) {
+    if (!n.is_batch_size_provider)  // these should go first
+      break;
+    bsps.emplace_back(&n, dynamic_cast<BatchSizeProvider *>(n.op.get()));
+    assert(bsps.back().second);
+  }
+  if (bsps.empty())
+    return nullptr;
+
+  return tasking::Task::Create([bsps, iter = iter_data, max_batch_size]() {
+    std::optional<int> bs;
+    assert(!bsps.empty());
+    for (auto [n, bsp] : bsps) {
+      int op_bs = bsp->NextBatchSize();
+      if (op_bs > max_batch_size) {
+        throw std::runtime_error(make_string(
+          "Batch too big! The input operator \"",
+          NodeName(*n),
+          "\" returned a batch size ", op_bs,
+          " which is larger than 'max_batch_size' for the pipeline."));
+      }
+      if (bs && *bs != op_bs)
+        throw std::runtime_error(make_string(
+          "Batch size clash! The input operator \"",
+          NodeName(*n),
+          "\" returned a batch size ", op_bs,
+          " which is different than ", *bs,
+          " retruned by \"", NodeName(*bsps[0].first), "\"."));
+      bs = op_bs;
+    }
+    assert(bs.has_value());
+    iter->default_batch_size = *bs;
+    for (auto &bsp : bsps)
+      bsp.second->Advance();
+  });
+}
+
+
+tasking::TaskFuture ExecGraph::Launch(tasking::Scheduler &sched) {
+  Sort();
+  Validate();
+  Analyze();
+
+  if (infer_batch_size_task_)
+    sched.AddSilentTask(infer_batch_size_task_);
+
+  std::optional<tasking::TaskFuture> ret;
+  for (auto &n : nodes_) {
+    auto maybe_future = n.Launch(sched);
```

Comment:
Did you consider adding a full scheduled graph/iteration to the scheduler/executor for execution? Or treating it as a some kind of transaction? I wonder what is the impact of adding more work when the executor is already running? Do you have any thoughts on the performance?

Reply:
1. That could be achieved by adding the graph backwards.
2. The "tranasction" semantics is more or less maintained as long as tasks don't wait for external conditions (unfortunately, that can happen with blocking ES, but there's hope)
3. I decided for this order of adding work because there would be _fewer_ thread switches - if a worker thread already has work to do, then it's busy - it doesn't wait, there are no futile notifications. We just add the work to the scheduler and we won't compete with the worker threads for locks.
4. E2E time of autoaugment tests showed minimal but measureable improvement (0.5%?)